[6기 박소현] TodoList with Ajax

pje1740 opened the pull request:
### 🎯 요구사항
- [x] User 추가하기
- [x] User의 투두리스트 불러오기
- [x] User 삭제하기
- [x] todoItem 추가하기
- [x] todoItem 불러오기
- [x] todoItem complete하기
- [x] todoItem 삭제하기
- [x] todoItem contents 내용 수정하기

### 🎯🎯 심화 요구사항
- [x] 데이터를 불러오기전 로딩바를 이용해, 사용자가 데이터가 불러와지고 있다는 것을 보여줍니다.
- [x] fetch api 사용하는 부분을 async await을 사용하여 리팩토링합니다.
- [x] github issue에서 라벨을 붙이는 것처럼, 우선순위에 따라서 badge를 추가합니다.
- [x] ES6 impot & export를 이용해 자바스크립트 파일을 리팩토링합니다.

### 코멘트
상태 관리 구현을 해보려고 고민을 하다가 이번주도 PR이 많이 늦었습니다 ㅠㅠ 리팩토링은 사치였습니다...
이번에 중점적으로 구현해보고자 했던 것은 createStore를 통한 특정 상태를 만들고, 해당 상태에 변화를 주는 이벤트가 발생하면, 그 이벤트의 변화를 감지하면 작동해야 하는 함수를 등록하여 변화와 동시에 자동으로 해당 함수들이 실행되게 하는 것이었습니다. 옵저버 패턴을 제 입맛대로 작성해보았는데, 비효율적으로 돌아가는 부분이 있다면 말씀 주세요! 아래는 대략적인 함수 설명입니다.

store.subscribe - 첫번째 인자로 넘겨준 상태 변화 이벤트가 발생하면 호출할 함수를 두번째 인자로 넘겨줍니다. 이벤트 구독입니다.
dispatch - 이벤트명과 필요한 데이터를 payload로 넘겨줍니다.

첫 주에 엉성하게 구현했던 옵저버 패턴을 싹 갈아엎으면서 많이 배운 거 같습니다. 
로딩바는 api 쪽에 급하게 넣어뒀는데, 어떻게 관리하는 게 좋았을까요...? 🤔

pje1740 committed:
[Chore] :package: apply basic settings

- eslintrc
- prettierrc
- npm
- folder reorganization
- gitignore

pje1740 committed:
[Feat] :zap:️ link index.js to index.html as module

pje1740 committed:
[Feat] :zap:️ upload to github

yujo11 committed:
feat: User를 추가하는 기능

pje1740 committed:
[Feat] :zap:️ upload upto get users

yujo11 committed:
feat: event delegation for user create button

pje1740 committed:
[Feat] :zap:️ show alert when wrong format of username entered

pje1740 committed:
[Feat] :zap:️ add return error for apis

pje1740 committed:
[Feat] :zap:️ create Store.js for creating states

state 및 관련 액션을 처리하기 위한 리듀서를 묶어서 관리할 수 있게 해주는 함수 생성

pje1740 committed:
[Feat] :zap:️ create users store

상단 유저 목록 데이터를 관리하는 store생성

pje1740 committed:
[Refactor] :rainbow: refactor UserList component

state를 이용해 옵저버 패턴으로 view 렌더링
클래스형에서 함수형으로 수정

pje1740 committed:
[Feat] :zap:️ add user remove button

pje1740 committed:
[Chore] :package: add no-underscore-dangle off option

pje1740 committed:
[Feat] :zap:️ add getUser api call

pje1740 committed:
[Refactor] :rainbow: fix variable names to fit the template

여러 상태에서 변수명 수정을 최소화하기 위해 변수명 수정

pje1740 committed:
[Feat] :zap:️ add store for selectedUser

선택된 유저를 상태로 관리

pje1740 committed:
[Feat] :zap:️ highlight selected user and connect selectedUser state

선택한 유저에 따라 view를 렌더. 상태와 연결

pje1740 committed:
[Feat] :zap:️ add user deletion from UserList

pje1740 committed:
[Feat] :zap:️ add component for page title

pje1740 committed:
[Feat] :zap:️ create Todolist component to display todo items

pje1740 committed:
[Feat] :zap:️ add delete Todo item function

pje1740 committed:
[Feat] :zap:️ add todo item contents edit function

pje1740 committed:
[Feat] :zap:️ add toggle complete function

pje1740 committed:
[Feat] :zap:️ add editing priority function

pje1740 committed:
[Feat] :zap:️ add loading for get User

JunseokPark13 commented on src/api/api.js:
```diff
@@ -0,0 +1,216 @@
+import Loading from '../components/Loading.js';
+
+const BASE_URL = 'https://js-todo-list-9ca3a.df.r.appspot.com';
+const ADD_USER = '/api/users';
+const GET_USERS = '/api/users';
+const GET_USER = '/api/users/';
+const DELETE_USER = '/api/users/';
+const GET_TODO = (userId) => `/api/users/${userId}/items`;
+const ADD_TODO = (userId) => `/api/users/${userId}/items`;
+const DELETE_TODO = (userId, itemId) => `/api/users/${userId}/items/${itemId}`;
+const EDIT_TODO_CONTENTS = (userId, itemId) => `/api/users/${userId}/items/${itemId}`;
+const TOGGLE_COMPLETE = (userId, itemId) => `/api/users/${userId}/items/${itemId}/toggle`;
+const DELETE_ALL = (userId) => `/api/users/${userId}/items`;
+const EDIT_PRIORITY = (userId, itemId) => `/api/users/${userId}/items/${itemId}/priority`;
```
저는 조건별로 url 를 따로 조립해주는 것만 생각했는데, 이런 식으로 이름을 부여해두는 것이 훨씬 직관적이고 다른 사람들이 알아보기도 편한 것 같습니다 👍

yujo11 replied:
API를 이렇게 명세하는것도 좋네요!👍👍👍

JunseokPark13 commented on src/components/UserDeleteButton.js:
```diff
@@ -0,0 +1,7 @@
+class UserRemoveButton {
+  static render() {
+    return `<button class="ripple user-delete-button">선택 유저 삭제</button>`;
+  }
+}
```
대부분의 분들이 이렇게 유저 삭제 버튼을 만드셨군요.. 저도 이렇게 해볼걸 그랬습니다 ㅠ

JunseokPark13 reviewed (approved):
코드 잘 봤습니다! 다들 코드를 모듈화하여 분리하는걸 엄청 능숙하게 잘하시는군요.. 이런 모듈화에 익숙하지 않은 저에게 많은 참고가 되었습니다. 감사합니다!!

yujo11 commented on src/api/api.js:
```diff
@@ -0,0 +1,216 @@
+const renderLoading = Loading().render;
+
+const addUser = async (name) => {
+  try {
+    const response = await fetch(`${BASE_URL}${ADD_USER}`, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({ name }),
+    });
+
+    if (!response.ok) {
+      throw new Error(response.status);
+    }
+    return response.json();
+  } catch (err) {
+    alert('addUser fail');
+    return err;
+  }
```
중복되는 ```try - catch``` 구문을 함수로 분리하면 좋을거 같습니다!

yujo11 commented on src/components/Title.js:
```diff
@@ -0,0 +1,16 @@
+import selectedUserStore, { GET_USER } from '../modules/selectedUser.js';
+
+const Title = () => {
+  const $title = document.querySelector('h1#user-title');
```
```suggestion
  const $title = document.querySelector('#user-title');
```

CSS id속성의 경우 한 문서에 고유한 값 하나만 존재할 수 있기 때문에 위 코드처럼 아이디만 명시해서 바로 파싱해오곤 했는데 태그 이름을 명시해주면 해당 DOM 요소가 어떤 태그인지도 알 수 있겠네요! 좋은 방법인거 같습니다👍👍👍

yujo11 commented on src/components/Todolist.js:
```diff
@@ -0,0 +1,126 @@
+
+  const render = () => {
+    const todos = selectedUserStore.getState().todoList;
+    $todolist.innerHTML = todos.map((todo) => todoItemRender(todo)).join('\n');
```
```join()```메서드를 사용할 때 개행을 넣어주는 부분이 센스있네요! 좋은 방법 같습니다!!👍👍👍

yujo11 commented on src/components/Todolist.js:
```diff
@@ -0,0 +1,126 @@
+    $todolist.innerHTML = todos.map((todo) => todoItemRender(todo)).join('\n');
+  };
+
+  const onDeleteTodo = ({ target }) => {
```
혹시 이 부분만 구조 분해를 통해 ```target```을 받아오신 이유가 있을까요?

pje1740 replied:
아마 통일성이 부족해보이는 건 리팩토링을 거치지 못해서 그럴 거 같습니다...ㅠㅠ 어떤 경우에는 이벤트 객체를 접근해야 하는 일이 있어서 그냥 event를 받아오고, 어떤 경우는 필요가 없어서 바로 target으로 받아왔는데 한번 정리를 해야겠네요....!

yujo11 commented on src/components/UserList.js:
```diff
@@ -0,0 +1,69 @@
+        return;
+      }
+      if (userName.trim().length < MINIMUN_USER_LENGTH) {
+        alert('닉네임은 2자 이상 입력해야 합니다.');
```
```suggestion
        alert(`닉네임은 ${MINIMUN_USER_LENGTH}자 이상 입력해야 합니다.`);
```

상수로 ```MINIMUN_USER_LENGTH```를 분리하고 조건 역시 ```MINIMUN_USER_LENGTH```를 사용해서 검증하는만큼 alert에도 ```MINIMUN_USER_LENGTH```를 넣어서 같이 관리하게 되면 좋을거 같습니다!

pje1740 replied:
오 좋은 아이디어 감사합니다~

yujo11 commented on js/index.js:
```diff
@@ -1,7 +1,6 @@
-}
-
-const userCreateButton = document.querySelector('.user-create-button')
-userCreateButton.addEventListener('click', onUserCreateHandler)
+  const userName = prompt('추가하고 싶은 이름을 입력해주세요.');
+};
 
+const userCreateButton = document.querySelector('.user-create-button');
+userCreateButton.addEventListener('click', onUserCreateHandler);
```
사용하지 않는 파일은 지워주시면 좋을거 같습니다!

yujo11 reviewed (approved):
코드 잘 봤습니다! Store와 dispatch를 사용해서 관리한게 인상 깊었네요!

전체적인 depth가 깊지 않고 변수명도 명확해서 가독성이 좋다고 느꼈습니다! 고생하셨습니다!!👍👍👍

pje1740 committed:
[Refactor] :rainbow: refactor code

- 불필요한 파일 삭제
- api 인자 비구조 할당으로 받아오도록 통일
- 이벤트 핸들러 가능한 경우 target만 비구조 할당으로 받아오도록 통일

domuk-k commented on src/store/Store.js:
```diff
@@ -0,0 +1,72 @@
+  };
+};
+
+export default { createStore, createAsyncStore };
```
옵저버 패턴의 코드를 처음보는 데, 흥미롭네요! 깔끔해서 잘 읽히는 느낌입니다! pull해봐야겠어요ㅎㅎ
감사합니다!

skid901 commented on src/App.js:
```diff
@@ -0,0 +1,18 @@
+import UserItem from './components/UserItem.js';
+import UserList from './components/UserList.js';
+import Title from './components/Title.js';
+import Todolist from './components/Todolist.js';
+import TodoInput from './components/TodoInput.js';
+import TodolistFooter from './components/TodolistFooter.js';
```
`components` 폴더에 `index.js` 파일을 만들고,
외부에 제공할 모든 컴포넌트를 `index.js` 파일에서 `export from` 구문으로 제공하면
조금 더 편리한 인터페이스를 제공할 수 있을거 같아요.

```js
// src/components/index.js
export { default as UserItem } from './UserItem.js';
export { default as UserList } from './UserList.js';

// src/App.js
import { UserItem, UserList } from './components/index.js';
```

pje1740 replied:
오 이렇게 작성하는 건 처음 보는 거 같아요!! 말씀 주신 방향으로 한번 고쳐보겠습니다~

jiseonk replied:
오 저도 한 수 배우고 갑니다!

skid901 commented on src/api/api.js:
```diff
@@ -0,0 +1,216 @@
+import Loading from '../components/Loading.js';
+
+const BASE_URL = 'https://js-todo-list-9ca3a.df.r.appspot.com';
+const ADD_USER = '/api/users';
+const GET_USERS = '/api/users';
+const GET_USER = '/api/users/';
+const DELETE_USER = '/api/users/';
+const GET_TODO = (userId) => `/api/users/${userId}/items`;
+const ADD_TODO = (userId) => `/api/users/${userId}/items`;
+const DELETE_TODO = (userId, itemId) => `/api/users/${userId}/items/${itemId}`;
```
`url` 경로를 제공해주는 변수들의 데이터 타입이 문자열과 함수로 혼재되어 있어서,
외부에서 사용하는 입장에서 헷갈릴 수 있을거 같아요.
동일한 역할을 수행하는 변수 또는 모듈 간에는 동일한 인터페이스를 제공해주시면 좋을 것 같아요.

```js
const GET_USERS = () => '/api/users';
const GET_TODO = (userId) => `/api/users/${userId}/items`;
```

pje1740 replied:
이것도 리팩토링 요소로 생각 중이었다가 완전 잊고있었네요ㅠㅠㅠ 찾아주셔서 감사합니다!

skid901 commented on src/components/Todolist.js:
```diff
@@ -0,0 +1,126 @@
+const Todolist = () => {
+  const $todolist = document.querySelector('ul.todo-list');
+  const { render: todoItemRender } = TodoItem();
+  const priorities = {
+    0: 'NONE',
+    1: 'FIRST',
+    2: 'SECOND',
+  };
```
`index`로 `property`를 제공하는 객체는 배열로 대체하셔도 괜찮을 거 같아요.

```js
const priorities = [ 'NONE', 'FIRST', 'SECOND' ];
```

skid901 commented on src/components/Todolist.js:
```diff
@@ -0,0 +1,126 @@
+    const inputEdit = closestLi.querySelector('input.edit');
+    inputEdit.style.display = 'block';
+    inputEdit.focus();
+    // 캐럿 위치 끝으로 이동
+    inputEdit.setSelectionRange(inputEdit.value.length, inputEdit.value.length);
```
사용자 친화적인 기능 구현 배워갑니다 ^^7
저도 적용해 봐야겠네요!!

skid901 commented on src/components/Todolist.js:
```diff
@@ -0,0 +1,126 @@
+        switchBackToView();
+        e.target.value = closestLi.querySelector('label span').innerText;
+      }
+    }
```
1주차 미션 피드백에서 `빠른 실패` 항목을 적용해 보시면,
코드 인텐트를 줄여 좀더 가독성이 좋은 코드가 될 것 같아요.

skid901 reviewed (approved):
디자인 패턴을 멋지게 적용하신 것 같아요!!
덕분에 리뷰하면서 많은 공부가 되었습니다 ^^7
피드백을 조금 남겨보았습니다. 여유 되실 때 한번 봐주시면 좋을 것 같아요 ㅎㅎ

pje1740 committed:
[Refactor] :rainbow: refactor api, Todolist code

- 빠른 리턴 추가
- api url 변수에서 모두 생성 함수로 변경
- store의 테스트용 console.log

DongHwani commented on src/api/api.js:
```diff
@@ -8,6 +8,8 @@ const ADD_TODO = (userId) => `/api/users/${userId}/items`;
 const EDIT_TODO_CONTENTS = (userId, itemId) => `/api/users/${userId}/items/${itemId}`;
 const TOGGLE_COMPLETE = (userId, itemId) => `/api/users/${userId}/items/${itemId}/toggle`;
+const DELETE_ALL = (userId) => `/api/users/${userId}/items`;
+const EDIT_PRIORITY = (userId, itemId) => `/api/users/${userId}/items/${itemId}/priority`;
```
상수로 API관리하니깐 엄청 깔끔하네요.  배우고 갑니다 👍

DongHwani commented on src/api/api.js:
```diff
@@ -159,6 +161,40 @@ const toggleComplete = async ({ userId, itemId }) => {
+    });
+
+    if (!response.ok) {
+      throw new Error(response.status);
```
Error 객체를 사용한 예외처리방법이 있었네요. 저도 한번 적용해봐야겠습니다 :)

DongHwani commented on src/api/api.js:
```diff
@@ -123,6 +124,24 @@ const deleteTodo = async ({ userId, itemId }) => {
+    return response.json();
+  } catch (err) {
+    alert(err);
+    return err;
```
코드 전반적으로 err값을 리턴해주는 형식인데  
별도로 에러를 처리하는 곳이 있을까요 ?

pje1740 replied:
좋은 지적입니다... 관습적으로 던져놓고 별도로 처리하질 못했네요! 3주차에선 유념하면서 작업해봐야겠습니다! 감사합니다 :)

DongHwani reviewed (approved):
안녕하세요 소연님, 리뷰가 많이 늦어져서 죄송합니다 

세부적으로 깔끔하게 만드신 흔적들이 보이네요. 특히나 vuex처럼 만드신 구조를 보고
많이 배워갑니다 👍  :)

2주차 미션도 고생많으셨습니다!

jiseonk commented on src/components/TodolistFooter.js:
```diff
@@ -0,0 +1,71 @@
+  DELETE_ALL,
+} from '../modules/selectedUser.js';
+
+const TodolistFooter = () => {
+  const $todoCount = document.querySelector('span.todo-count strong');
+  const $showAllButton = document.querySelector('ul.filters a.all');
+  const $showActiveButton = document.querySelector('ul.filters a.active');
+  const $showCompleteButton = document.querySelector('ul.filters a.completed');
+  const $deleteAllButton = document.querySelector('button.clear-completed');
+
+  const renderTodoCount = () => {
+    const count = selectedUserStore.getState().todoList.length;
+    $todoCount.innerText = count;
+  };
+
+  const toggleFilterSelection = (buttonType) => {
```
투두리스트 하단 부분에 필요한 기능들을 TodolistFooter로 묶어서 정리해주신 게 좋은 아이디어 같아요! 저도 이렇게 코드 짜보고 싶네요

jiseonk commented on src/components/UserList.js:
```diff
@@ -0,0 +1,69 @@
+    }
+  };
+
+  const render = async () => {
+    const users = await usersStore.getState();
+    const userButtons = users.map((user) => UserItem.render(user._id, user.name));
+    userButtons.push(UserAddButton.render(), UserDeleteButton.render());
+    $userList.innerHTML = userButtons.join('\n');
+  };
```
렌더 함수를 참 깔끔하게 잘 쓰시는 거 같아요! 
렌더 함수를 어떻게 구현해야할지 잘 감이 안왔었는데 많이 배우고 갑니다.
그리고 다들 버튼을 추가하는 함수를 따로 만드셨네요. 저는 처음부터 html에 집어넣었는데 이 방식이 유저리스트를 수정할 때 더 깔끔하게 느껴지는 것 같고 좋네요👍👍

imakerjun commented on src/api/api.js:
```diff
@@ -0,0 +1,216 @@
+      method: 'GET',
+    });
+
+    if (!response.ok) {
+      throw new Error(response.status);
+    }
+    return response.json();
```
중복되는 부분일 것 같은데 재사용하면 더 좋을 것 같아요! 😄

imakerjun commented on src/components/Todolist.js:
```diff
@@ -0,0 +1,128 @@
+    const inputEdit = closestLi.querySelector('input.edit');
+    inputEdit.style.display = 'block';
+    inputEdit.focus();
+    // 캐럿 위치 끝으로 이동
```
주석보다는 메서드 네이밍으로 표현할 수 있으면 더 좋은 코드가 될 것 같아요!

imakerjun commented on src/components/Todolist.js:
```diff
@@ -0,0 +1,128 @@
+  $todolist.addEventListener('click', onDeleteTodo);
+  $todolist.addEventListener('dblclick', onOpenTodoEdit);
+  $todolist.addEventListener('keyup', onKeyupTodoEdit);
+  $todolist.addEventListener('click', onToggleComplete);
```
111 라인과 똑같이 같은 엘리먼트에 클릭 이벤트를 추가하고 있는데, 클릭이벤트는 한번만 할당하고, 콜백함수에서 정확히 어떤 엘리먼트가 위임받았는지 분기하는게 더 좋을 것 같아요! 그렇지 않으면 이벤트 핸들러가 불필요하게 매번 2번씩 작동되야하는 것 같아요 😭

imakerjun commented on src/components/TodolistFooter.js:
```diff
@@ -0,0 +1,71 @@
+    [$showAllButton, $showActiveButton, $showCompleteButton].map((btn) =>
+      btn.classList.remove('selected'),
+    );
+    buttonType === 'all' ? $showAllButton.classList.add('selected') : null;
```
반드시 null을 넣어야 하는 경우가 아니라면 아래와 같이 간략하게 작성하는 것도 가능합니다!
`buttonType === 'all' && $showAllButton.classList.add('selected')`

pje1740 replied:
아 &&를 사용해서 이렇게 대체해서 적을 수 있는 거였군요! 이제 정확히 이해했습니다! 감사합니다!

imakerjun commented on src/components/UserList.js:
```diff
@@ -0,0 +1,69 @@
+  const onAddUser = async ({ target }) => {
+    if (target.classList.contains('user-create-button')) {
+      const name = prompt('이름을 입력해주세요');
+      if (name === null) {
```
`if (!!name)`으로 비교해주시면 null 뿐만 아니라 undefiend나, 빈문자열인 경우도 체크할 수 있어요 😄

imakerjun commented on src/components/UserList.js:
```diff
@@ -0,0 +1,69 @@
+  const $userList = document.querySelector('#user-list');
+
+  const onAddUser = async ({ target }) => {
+    if (target.classList.contains('user-create-button')) {
```
1주차 공통피드백으로 드린 빠른 실패를 적용해보셔도 좋을 것 같아요!
